Reference implementations for softmax, log_softmax, logsumexp

IvanYashchuk opened the pull request:
This PR adds references for:

- `torch.softmax`
- `torch.log_softmax`
- `torch.logsumexp`

Unfortunately, none of them currently pass `test_python_ref_executor` even with `"aten"` executor.

IvanYashchuk committed:
Add reference implementation for softmax

IvanYashchuk committed:
Add reference implementation for logsumexp

IvanYashchuk committed:
Add reference implementation for log_softmax

ngimel commented:
What are the errors with aten executor?

mruberry commented on torch/_refs/__init__.py:
```diff
@@ -520,6 +520,50 @@ def log10(a):
+) -> TensorLikeType:
+    result_dtype = dtype or a.dtype
+    computation_dtype = utils.get_computation_dtype(a.dtype)
+    a = prims.convert_element_type(a, computation_dtype)
```
Let's make this conditional on the conversion being required -- we actually have a function for this if you'd prefer, but it's in an awkward place (it could be moved to utils if you'd rather use it vs. a custom conditional) 

https://github.com/pytorch/pytorch/blob/38e717dc876e0f9af66b32e9cc59b44b1f9edf27/torch/_prims/wrappers.py#L20

IvanYashchuk replied:
Thanks! `_maybe_convert_to_dtype` is already used in this file, so we can keep it defined in the awkward place for now 🙂

mruberry commented on torch/_refs/__init__.py:
```diff
@@ -520,6 +520,50 @@ def log10(a):
+    computation_dtype = utils.get_computation_dtype(a.dtype)
+    a = prims.convert_element_type(a, computation_dtype)
+    a_max = amax(a, dim, keepdim=True)
+    shifted = a - a_max
```
Nit: comment for why the shift occurs would be nice

IvanYashchuk replied:
Oh, the shift is actually not required here because stabilized logsumexp is used. I will remove it.

mruberry commented on torch/_refs/__init__.py:
```diff
@@ -520,6 +520,50 @@ def log10(a):
+    a_max = amax(a, dim, keepdim=True)
+    shifted = a - a_max
+    shifted_logsumexp = logsumexp(shifted, dim, keepdim=True)
+    return prims.convert_element_type(
```
We should also make this conditional on the conversion being required

mruberry commented on torch/_refs/__init__.py:
```diff
@@ -520,6 +520,50 @@ def log10(a):
+    )
+
+
+def _squeeze_multiple(a: TensorLikeType, dims: DimsSequenceType) -> TensorLikeType:
```
Unify this with prims.squeeze -- see 

https://github.com/pytorch/pytorch/blob/38e717dc876e0f9af66b32e9cc59b44b1f9edf27/torch/_prims/__init__.py#L1636

I think we only need one? But maybe this implementation is better for the prim?

IvanYashchuk replied:
I didn't notice that `prims.squeeze` works for multiple specified dimensions. Both implementations are fine, I will not touch `prims.squeeze` at this time.

mruberry commented on torch/_refs/__init__.py:
```diff
@@ -520,6 +520,50 @@ def log10(a):
+@out_wrapper
+def logsumexp(
+    a: TensorLikeType,
+    dims: DimsType,
```
"dim"? 

https://pytorch.org/docs/master/generated/torch.logsumexp.html?highlight=logsumexp#torch.logsumexp

IvanYashchuk replied:
Right, it should be dim to match the torch namespace. I think I was confused because it also accepts and works for several dimensions
> dim (int or tuple of python:ints)
```py
import torch
a = torch.ones(3, 3)
torch.logsumexp(a, (0, 1))
# tensor(3.1972)
```

mruberry commented on torch/_refs/__init__.py:
```diff
@@ -520,6 +520,50 @@ def log10(a):
+    keepdim: bool = False,
+) -> TensorLikeType:
+    dims = utils.canonicalize_dims(a.ndim, dims)
+    # ATen specifies int[1] type dims which expands integers to tuples of length 1
```
Nice comment

mruberry commented on torch/_refs/__init__.py:
```diff
@@ -520,6 +520,50 @@ def log10(a):
+        a_max_squeezed = _squeeze_multiple(a_max, dims) if not keepdim else a_max
+        result = log(sum(exp(a - a_max), dims, keepdim=keepdim)) + a_max_squeezed
+    else:
+        result = log(sum(exp(a), dims, keepdim=keepdim))
```
Add a comment for what this case covers (integer and boolean dtypes)

mruberry commented on torch/_refs/__init__.py:
```diff
@@ -520,6 +520,50 @@ def log10(a):
+        dims = (dims,)
+    if utils.is_float_dtype(a.dtype) or utils.is_complex_dtype(a.dtype):
+        a_max = amax(a, dims, keepdim=True)
+        a_max = where(abs(a_max) == float("inf"), 0.0, a_max)
```
really elegant code here

mruberry commented on torch/_refs/__init__.py:
```diff
@@ -2017,6 +2061,23 @@ def stack(tensors: TensorSequenceType, dim: int = 0) -> TensorLikeType:
+) -> TensorLikeType:
+    result_dtype = dtype or a.dtype
+    computation_dtype = utils.get_computation_dtype(a.dtype)
+    a = prims.convert_element_type(a, computation_dtype)
```
Similar comments here as with log_softmax re: conditional conversions

mruberry reviewed (approved):
This is awesome, @IvanYashchuk! I made some inline comments for your review, nothing major, and the lint job needs to be fixed, but approving this for velocity because I'm sure you'll sort out the review and the jobs.

IvanYashchuk committed:
Black formatting

IvanYashchuk committed:
Add multiple dims input for logsumexp

IvanYashchuk committed:
Shift is not required in log_softmax because stabilized logsumexp is used

IvanYashchuk committed:
Use _maybe_convert_to_dtype

IvanYashchuk committed:
Add comments for logsumexp

IvanYashchuk committed:
Use prims.squeeze instead of _squeeze_multiple

IvanYashchuk committed:
Rename dims -> dim

IvanYashchuk committed:
Skip MyPy return value of _maybe_convert_to_dtype

IvanYashchuk committed:
Fix mypy assignment error for softmax

IvanYashchuk committed:
Fix mypy assignment error for log_softmax

IvanYashchuk committed:
Fix mypy arg-type error in softmax

IvanYashchuk committed:
Fix mypy arg-type error for amax in logsumexp

IvanYashchuk commented:
`_maybe_convert_to_dtype` is now used for optional conversions. I removed unnecessary shift from log_softmax and fixed mypy errors. In case when `_maybe_convert_to_dtype` is the returned variable mypy errors are silenced.

IvanYashchuk committed:
Use gradcheck_fast_mode=False for logsumexp

Forward AD test fails with new sample input that reduces over multiple
dims

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/viable/strict' into refs-softmax

IvanYashchuk commented:
@pytorchbot merge -g
